// pnotify/lib/es/PNotifyCallbacks.js
/* src/PNotifyCallbacks.html generated by Svelte v2.16.1 */
import PNotify from "./PNotify.js";

let _open = PNotify.prototype.open;
let _close = PNotify.prototype.close;

const callbacks = (notice, options, name) => {
  let modules = notice ? notice.get().modules : options.modules;
  let cbs = (modules && modules.Callbacks) ? modules.Callbacks : {};
  return cbs[name] ? cbs[name] : () => true;
};

PNotify.prototype.open = function (...args) {
  let ret = callbacks(this, null, 'beforeOpen')(this);
  if (ret !== false) {
    _open.apply(this, args);
    callbacks(this, null, 'afterOpen')(this);
  }
};

PNotify.prototype.close = function (timerHide, ...args) {
  let ret = callbacks(this, null, 'beforeClose')(this, timerHide);
  if (ret !== false) {
    _close.apply(this, [timerHide, ...args]);
    callbacks(this, null, 'afterClose')(this, timerHide);
  }
};

function setup(Component) {
  Component.key = 'Callbacks';

  Component.getCallbacks = callbacks;

  let _alert = PNotify.alert;
  let _notice = PNotify.notice;
  let _info = PNotify.info;
  let _success = PNotify.success;
  let _error = PNotify.error;

  let init = (original, options) => {
    callbacks(null, options, 'beforeInit')(options);
    let notice = original(options);
    callbacks(notice, null, 'afterInit')(notice);
    return notice;
  };

  PNotify.alert = (options) => {
    return init(_alert, options);
  };
  PNotify.notice = (options) => {
    return init(_notice, options);
  };
  PNotify.info = (options) => {
    return init(_info, options);
  };
  PNotify.success = (options) => {
    return init(_success, options);
  };
  PNotify.error = (options) => {
    return init(_error, options);
  };

  // Register the module with PNotify.
  PNotify.modules.Callbacks = Component;
};

function create_main_fragment(component, ctx) {

	return {
		c: noop,

		m: noop,

		p: noop,

		d: noop
	};
}

function PNotifyCallbacks(options) {
	init(this, options);
	this._state = assign({}, options.data);
	this._intro = true;

	this._fragment = create_main_fragment(this, this._state);

	if (options.target) {
		this._fragment.c();
		this._mount(options.target, options.anchor);
	}
}

assign(PNotifyCallbacks.prototype, {
 	destroy: destroy,
 	get: get,
 	fire: fire,
 	on: on,
 	set: set,
 	_set: _set,
 	_stage: _stage,
 	_mount: _mount,
 	_differs: _differs
 });

PNotifyCallbacks.prototype._recompute = noop;

setup(PNotifyCallbacks);

function noop() {}

function init(component, options) {
	component._handlers = blankObject();
	component._slots = blankObject();
	component._bind = options._bind;
	component._staged = {};

	component.options = options;
	component.root = options.root || component;
	component.store = options.store || component.root.store;

	if (!options.root) {
		component._beforecreate = [];
		component._oncreate = [];
		component._aftercreate = [];
	}
}

function assign(tar, src) {
	for (var k in src) tar[k] = src[k];
	return tar;
}

function destroy(detach) {
	this.destroy = noop;
	this.fire('destroy');
	this.set = noop;

	this._fragment.d(detach !== false);
	this._fragment = null;
	this._state = {};
}

function get() {
	return this._state;
}

function fire(eventName, data) {
	var handlers =
		eventName in this._handlers && this._handlers[eventName].slice();
	if (!handlers) return;

	for (var i = 0; i < handlers.length; i += 1) {
		var handler = handlers[i];

		if (!handler.__calling) {
			try {
				handler.__calling = true;
				handler.call(this, data);
			} finally {
				handler.__calling = false;
			}
		}
	}
}

function on(eventName, handler) {
	var handlers = this._handlers[eventName] || (this._handlers[eventName] = []);
	handlers.push(handler);

	return {
		cancel: function() {
			var index = handlers.indexOf(handler);
			if (~index) handlers.splice(index, 1);
		}
	};
}

function set(newState) {
	this._set(assign({}, newState));
	if (this.root._lock) return;
	flush(this.root);
}

function _set(newState) {
	var oldState = this._state,
		changed = {},
		dirty = false;

	newState = assign(this._staged, newState);
	this._staged = {};

	for (var key in newState) {
		if (this._differs(newState[key], oldState[key])) changed[key] = dirty = true;
	}
	if (!dirty) return;

	this._state = assign(assign({}, oldState), newState);
	this._recompute(changed, this._state);
	if (this._bind) this._bind(changed, this._state);

	if (this._fragment) {
		this.fire("state", { changed: changed, current: this._state, previous: oldState });
		this._fragment.p(changed, this._state);
		this.fire("update", { changed: changed, current: this._state, previous: oldState });
	}
}

function _stage(newState) {
	assign(this._staged, newState);
}

function _mount(target, anchor) {
	this._fragment[this._fragment.i ? 'i' : 'm'](target, anchor || null);
}

function _differs(a, b) {
	return a != a ? b == b : a !== b || ((a && typeof a === 'object') || typeof a === 'function');
}

function blankObject() {
	return Object.create(null);
}

function flush(component) {
	component._lock = true;
	callAll(component._beforecreate);
	callAll(component._oncreate);
	callAll(component._aftercreate);
	component._lock = false;
}

function callAll(fns) {
	while (fns && fns.length) fns.shift()();
}
export default PNotifyCallbacks;
